develop/DSI/DSIOpenConfig/DSI_OpenConfig_Middleware.md

```diff
@@ -0,0 +1,456 @@
+
+### Accessing a specific instance
+
+If you need to access a specific instance in a `container`, you can use [ ] to specify the instance. For example: `interfaces/interface[name:Ethernet1]/state`
+
+Using this path will result in only reading or writing the *Ethernet1* instance of the *interfaces/interface/state* `container`.
+
+### Troubleshooting
+
+The errors that occur in the middleware will be logged by the `ILogger` object that gets passed along at construction of the [GnmiClient](xref:Skyline.DataMiner.Helper.OpenConfig.Api.GnmiClient).
+
+In case of a DataMiner connector, that will be the logging of the element. There is an extensive list of [error codes](xref:CommunicationGateway_ErrorCodes) available to help you troubleshoot.
+
+### DataMapper
+
+The [DataMinerConnectorDataMapper](xref:Skyline.DataMiner.Helper.OpenConfig.DataMapper.DataMinerConnectorDataMapper) is an object that sits between your device and the DataMiner connector. It will automatically parse the incoming notifications and populate DataMiner parameters or tables with the data.
+
+#### Map a YANG path to a parameter
+
+```csharp
+IDataMapper dataMapper = new DataMinerConnectorDataMapper(
+    protocol,
+    new List<IDataEntity>
+    {
+        new DataMinerConnectorParameter("system/memory/state/physical", Parameter.systemmemorystatephysical_400),
+        new DataMinerConnectorParameter("system/memory/state/reserved", Parameter.systemmemorystatereserved_404)
+    });
+```
+
+Having this configured as soon as a gNMI notification will come in with data for either the *system/memory/state/physical* or the *system/memory/state/reserved* YANG path, it will automatically do the set on the configured parameters. You need to pass it the `SLProtocol` object so it can execute those sets.
+
+It works perfectly for DataMiner tables as well, but the configuration is slightly different.
+
+```csharp
+IDataMapper dataMapper = new DataMinerConnectorDataMapper(
+    protocol,
+    new List<IDataEntity>
+    {
+        new DataMinerConnectorDataGrid("interfaces/interface/state", Parameter.Interfacesstate.tablePid, new List<IDataMinerConnectorDataGridColumn>
+        {
+            new DataMinerConnectorDataGridColumn("openconfig-interfaces:type", Parameter.Interfacesstate.Pid.interfacesstatetype),
+            new DataMinerConnectorDataGridColumn("openconfig-interfaces:mtu", Parameter.Interfacesstate.Pid.interfacesstatemtu),
+            new DataMinerConnectorDataGridColumn("openconfig-interfaces:loopback-mode", Parameter.Interfacesstate.Pid.interfacesstateloopbackmode)
+        }
+    });
+```
+
+> [!IMPORTANT]
+> When configuring the path for a column, always specify the YANG module name as well. Notifications of type `JSON` do not contain them but notifications of type `JSON_IETF` do. The DataMapper is capable of handling both, but for that reason the YANG module name needs to be known.
```

Comment:
What's "them" here? If it's the YANG module name, the plural doesn't really make sense given the singular "YANG module name" in the preceding sentence, so I'd suggest writing "it" instead.

Reply:
Changed to 'it'.